### PR TITLE
Happychat: Add better messaging when there are connectivity issues

### PR DESCRIPTION
**Pull request by mattwondra**
This combined with PR https://github.com/Automattic/wp-calypso/pull/12706 will fix issue https://github.com/Automattic/wp-calypso/issues/12183.

This PR adds a few Socket.IO lifecycle event dispatches that we can use to get more information on the status of a HC connection. This information is used to disable the text input and show a notice when certain connection issues are occurring, specifically:

- User tries to connect to HC but the server is not available
- User has an open connection and the HC server disconnects
- User has an open connection and their network goes down, causing a disconnect

For the last case, we may want to tweak the `pingTimeout` and `pingInterval` settings on the server. By default they are set to `60000ms` and `25000ms` respectively — which basically means it will take anywhere from 60–85s from a network connection failure until the user is notified of the failure. During this time the user may be firing off messages that will never reach the HE and won't show in their timeline, which is probably pretty frustrating.

#### HC is down on connecting
Point Calypso at your local HC but don't bring the service up. Go to http://calypso.localhost:3000/me/chat. You should see a notice `Connecting you with a Happiness Engineer…` followed quickly by `We're having trouble connecting to chat. Please bear with us while we try to reconnect…`. During this whole time the chat message box should be disabled.

Starting the local HC service should make the notice go away and HC should behave as expected.

#### HC service goes down after connecting
Point Calypso at your local HC with the service up. Go to http://calypso.localhost:3000/me/chat. You should see a notice `Connecting you with a Happiness Engineer…` followed by the notice disappearing. Start chatting with an operator.

Kill the local HC service. In Calypso you should immediately see a notice `We're having trouble connecting to chat. Please bear with us while we try to reconnect…`. The chat message box should be disabled.

Starting the HC service again should bring the chat back to normal status.

#### User's internet connection goes away
Point Calypso to staging HC. Go to http://calypso.localhost:3000/me/chat. You should see a notice `Connecting you with a Happiness Engineer…` followed by the notice disappearing. Start chatting with an operator.

Turn off your wifi. Wait for up to 90s and you should see a notice `We're having trouble connecting to chat. Please check your internet connection while we try to reconnect…`. The chat message box should be disabled.

Turning wifi on again should bring the chat back to normal status.

### CODE REVIEW

File: client/components/happychat/notices.jsx [hunk 1]
```diff
@@ -16,22 +16,29 @@ import {
...
 class Notices extends Component {
 	statusNotice() {
-		const { connectionStatus, chatStatus, translate } = this.props;
+		const { isServerReachable, connectionStatus, chatStatus, translate } = this.props;
+
+		if ( ! isServerReachable ) {
+			return translate( "We're having trouble connecting to chat. Please check your internet connection while we try to reconnect…" );
+		}
 
 		switch ( connectionStatus ) {
```
Comment (nosolosw) [inline client/components/happychat/notices.jsx]: We could use a shared constant to communicate [connectionStatus values](https://github.com/Automattic/wp-calypso/blob/eb428f37046b8e3dbaba7abe627287c176919eb1/client/state/happychat/reducer.js#L148) between this and the reducer.

File: client/state/happychat/constants.js [hunk 2]
```diff
@@ -1,2 +1,7 @@
+export const HAPPYCHAT_CONNECTION_ERROR_FORCED_CLOSE = 'forced close';
+export const HAPPYCHAT_CONNECTION_ERROR_PING_TIMEOUT = 'ping timeout';
+export const HAPPYCHAT_CONNECTION_ERROR_TRANSPORT_CLOSE = 'transport close';
+export const HAPPYCHAT_CONNECTION_ERROR_TRANSPORT_ERROR = 'transport error';
+
```
Comment (nosolosw) [inline client/state/happychat/constants.js]: What are these constants for? Apart from `HAPPYCHAT_CONNECTION_ERROR_PING_TIMEOUT` I haven't seen them in use anywhere.

File: client/components/happychat/notices.jsx (hunk 1 above)
Comment (mattwondra) [inline client/components/happychat/notices.jsx]: Thanks, this is on my to-do list before I move off the middleware cleanup tasks, but I didn't want to pollute existing PRs with that stuff.

File: client/state/happychat/constants.js (hunk 2 above)
Comment (mattwondra) [inline client/state/happychat/constants.js]: These come from the [Socket.IO client library](https://github.com/socketio/socket.io-client/blob/master/dist/socket.io.js) — search for calls to `onClose(` and the passed-in values are the possible reasons for the connection error.

I'll add a comment here to document this better.

**Comment by mattwondra** [comment]
> SocketIO may be maintaining the connection between local calypso and local HC because both run under localhost

Yep, I believe this is the case. :)

**Pull request merged**